File: solutions/clojure/robot-simulator/1/src/robot_simulator.clj
(ns robot-simulator)

(def deltas {:east {:x 1 :y 0} :north {:x 0 :y 1} :west {:x -1 :y 0}  :south {:x 0 :y -1}})

(defn robot[initial-coord initial-bearing]
    {:coordinates initial-coord :bearing initial-bearing}
)

(defn turn-left[bearing] 
         (condp = bearing 
            :east  :north
            :north  :west
            :west  :south
            :south :east
        )
)

(defn turn-right[bearing] 
        (condp = bearing 
            :east :south
            :north :east
            :west :north
            :south :west
        )
 )

(defn advance[robot]
        (let [delta (deltas (:bearing robot))]
            {:coordinates (merge-with + delta (:coordinates robot)) :bearing (:bearing robot)}       
        )
 )

(defn follow-letter[robot lt] 
      (condp = lt
        \A (advance robot) 
        \L {:coordinates (:coordinates robot) :bearing (turn-left (:bearing robot))  }
        \R {:coordinates (:coordinates robot) :bearing (turn-right (:bearing robot))  }
       )
)

(defn simulate[instructions robot]
       (reduce  follow-letter robot instructions)
)

File: solutions/dart/robot-simulator/1/lib/robot_simulator.dart
import 'package:robot_simulator/orientation.dart';
import 'package:robot_simulator/position.dart';


class Robot {

  Position position;
  Orientation orientation;

  Robot(this.position, this.orientation);


  void advance(){
    switch(orientation){
      case Orientation.east : position.x += 1;
      case Orientation.west : position.x -= 1;  
      case Orientation.north : position.y += 1;
      case Orientation.south : position.y -= 1;
    }
  }

  void turnLeft(){
    switch(orientation){
      case Orientation.north : orientation = Orientation.west;
      case Orientation.west :orientation = Orientation.south;
      case Orientation.south : orientation = Orientation.east;
      case Orientation.east : orientation = Orientation.north;
    }
  }

  void turnRight(){
    switch(orientation){
      case Orientation.north : orientation = Orientation.east;
      case Orientation.west : orientation = Orientation.north;
      case Orientation.south : orientation = Orientation.west;
      case Orientation.east : orientation = Orientation.south;
    }
  }

  
  void singleInstruction(String instruction){
    switch(instruction){
      case 'R' : turnRight();
      case 'L' : turnLeft();
      case 'A' : advance();
      default : throw ArgumentError('invalid instruction');        }    
  }
  
  void move(String instructions){
    for(var s in instructions.split('')){
      singleInstruction(s);
    }    
  }
  
}
